File: src/audio_stegano_app/lib/app/logistic_param_bounds.dart
/// UI and algorithm bounds for logistic map parameters (MATLAB port).
class LogisticParamBounds {
  LogisticParamBounds._();

  /// Slider / manual entry range for [r] (chaotic regime subset).
  static const double rMin = 3.5;
  static const double rMax = 4.0;

  /// Slider / manual entry range for [x0] (open interval (0,1)).
  static const double x0Min = 0.01;
  static const double x0Max = 0.99;

  static double clampR(double value) => value.clamp(rMin, rMax);
  static double clampX0(double value) => value.clamp(x0Min, x0Max);

  static double? tryParseR(String text) {
    final v = _tryParse(text);
    if (v == null || v <= 0 || v > 4.0) return null;
    return clampR(v);
  }

  static double? tryParseX0(String text) {
    final v = _tryParse(text);
    if (v == null || v <= 0 || v >= 1) return null;
    return clampX0(v);
  }

  static double? _tryParse(String text) {
    final t = text.trim().replaceAll(',', '.');
    if (t.isEmpty) return null;
    return double.tryParse(t);
  }
}

File: src/audio_stegano_desktop/src/AudioStegano.Desktop/LogisticParamBounds.cs
namespace AudioStegano.Desktop;

/// <summary>UI bounds for logistic map parameters (matches Flutter / MATLAB port).</summary>
public static class LogisticParamBounds
{
    public const double RMin = 3.5;
    public const double RMax = 4.0;
    public const double X0Min = 0.01;
    public const double X0Max = 0.99;

    public static double ClampR(double value) => Math.Clamp(value, RMin, RMax);
    public static double ClampX0(double value) => Math.Clamp(value, X0Min, X0Max);

    public static bool TryParseR(string text, out double result)
    {
        result = 0;
        if (!TryParse(text, out var v) || v <= 0 || v > 4.0) return false;
        result = ClampR(v);
        return true;
    }

    public static bool TryParseX0(string text, out double result)
    {
        result = 0;
        if (!TryParse(text, out var v) || v <= 0 || v >= 1.0) return false;
        result = ClampX0(v);
        return true;
    }

    private static bool TryParse(string text, out double value)
    {
        value = 0;
        var t = text.Trim().Replace(',', '.');
        return double.TryParse(t, System.Globalization.NumberStyles.Float,
            System.Globalization.CultureInfo.InvariantCulture, out value);
    }
}
